File: frontend/mobile_app/lib/data/checkout_repository.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../core/config/app_config.dart';
import '../models/cart.dart';
import '../models/order.dart';

/// A thrown checkout failure carrying a human-readable message.
class CheckoutException implements Exception {
  const CheckoutException(this.message);
  final String message;
  @override
  String toString() => message;
}

/// Optional customer capture attached to a checkout.
class CustomerDraft {
  const CustomerDraft({this.name, this.mobile});

  final String? name;
  final String? mobile;

  bool get isEmpty =>
      (name == null || name!.trim().isEmpty) &&
      (mobile == null || mobile!.trim().isEmpty);

  Map<String, dynamic> toJson() => <String, dynamic>{
    if (name != null && name!.trim().isNotEmpty) 'name': name!.trim(),
    if (mobile != null && mobile!.trim().isNotEmpty) 'mobile': mobile!.trim(),
  };
}

/// Persists the controller-side cart as an order via the backend
/// (`POST /api/cart/:sessionId/checkout`). Swapped for [MockCheckoutRepository]
/// in standalone mode so the offline demo still completes a sale.
abstract interface class CheckoutRepository {
  Future<Order> checkout({
    required String sessionId,
    required String token,
    required Cart cart,
    CustomerDraft? customer,
  });
}

/// Maps cart lines to the backend's checkout item shape.
List<Map<String, dynamic>> _itemsPayload(Cart cart) => cart.items
    .map(
      (CartItem i) => <String, dynamic>{
        'productId': i.product.id,
        'variantId': i.variantId,
        'size': i.size,
        'quantity': i.quantity,
      },
    )
    .toList();

class HttpCheckoutRepository implements CheckoutRepository {
  HttpCheckoutRepository({http.Client? client})
    : _client = client ?? http.Client();

  final http.Client _client;
  static const Duration _timeout = Duration(seconds: 8);

  @override
  Future<Order> checkout({
    required String sessionId,
    required String token,
    required Cart cart,
    CustomerDraft? customer,
  }) async {
    final Map<String, dynamic> body = <String, dynamic>{
      'items': _itemsPayload(cart),
      if (customer != null && !customer.isEmpty) 'customer': customer.toJson(),
    };
    late final http.Response res;
    try {
      res = await _client
          .post(
            AppConfig.http('/api/cart/$sessionId/checkout'),
            headers: <String, String>{
              'content-type': 'application/json',
              'authorization': 'Bearer $token',
            },
            body: jsonEncode(body),
          )
          .timeout(_timeout);
    } catch (_) {
      throw const CheckoutException(
        'Could not reach the server. Check the connection.',
      );
    }
    final Map<String, dynamic> json =
        jsonDecode(res.body) as Map<String, dynamic>;
    if (res.statusCode == 200 || res.statusCode == 201) {
      return Order.fromJson(json);
    }
    final Object? err = json['error'];
    final String message = err is Map<String, dynamic>
        ? (err['message'] as String? ?? 'Checkout failed')
        : 'Checkout failed';
    throw CheckoutException(message);
  }
}

/// Offline/demo checkout: builds an [Order] locally from the cart totals.
class MockCheckoutRepository implements CheckoutRepository {
  const MockCheckoutRepository();

  @override
  Future<Order> checkout({
    required String sessionId,
    required String token,
    required Cart cart,
    CustomerDraft? customer,
  }) async {
    return Order(
      id: 'demo_${DateTime.now().millisecondsSinceEpoch}',
      status: 'placed',
      itemCount: cart.count,
      subtotal: cart.subtotal,
      tax: cart.tax,
      total: cart.total,
      createdAt: DateTime.now().toIso8601String(),
      customerName: customer?.name,
    );
  }
}

File: frontend-rn/mobile_app/src/data/checkoutRepository.ts
import { AppConfig } from '../config/appConfig';
import { Cart, CartItem } from '../models/cart';
import { Order } from '../models/order';

/// A thrown checkout failure carrying a human-readable message.
export class CheckoutException extends Error {}

/// Optional customer capture attached to a checkout.
export class CustomerDraft {
  readonly name?: string;
  readonly mobile?: string;
  constructor(init?: { name?: string; mobile?: string }) {
    this.name = init?.name;
    this.mobile = init?.mobile;
  }
  get isEmpty(): boolean {
    return (this.name == null || this.name.trim().length === 0) && (this.mobile == null || this.mobile.trim().length === 0);
  }
  toJson(): Record<string, any> {
    const out: Record<string, any> = {};
    if (this.name != null && this.name.trim().length > 0) out.name = this.name.trim();
    if (this.mobile != null && this.mobile.trim().length > 0) out.mobile = this.mobile.trim();
    return out;
  }
}

export interface CheckoutRepository {
  checkout(opts: { sessionId: string; token: string; cart: Cart; customer?: CustomerDraft }): Promise<Order>;
}

function itemsPayload(cart: Cart): Record<string, any>[] {
  return cart.items.map((i: CartItem) => ({
    productId: i.product.id,
    variantId: i.variantId,
    size: i.size,
    quantity: i.quantity,
  }));
}

/// Persists the controller-side cart as an order via the backend. Ported 1:1.
export class HttpCheckoutRepository implements CheckoutRepository {
  private static readonly timeoutMs = 8000;

  async checkout(opts: { sessionId: string; token: string; cart: Cart; customer?: CustomerDraft }): Promise<Order> {
    const body: Record<string, any> = { items: itemsPayload(opts.cart) };
    if (opts.customer != null && !opts.customer.isEmpty) body.customer = opts.customer.toJson();

    const controller = new AbortController();
    const t = setTimeout(() => controller.abort(), HttpCheckoutRepository.timeoutMs);
    let res: Response;
    try {
      res = await fetch(AppConfig.http(`/api/cart/${opts.sessionId}/checkout`), {
        method: 'POST',
        headers: { 'content-type': 'application/json', authorization: `Bearer ${opts.token}` },
        body: JSON.stringify(body),
        signal: controller.signal,
      });
    } catch {
      throw new CheckoutException('Could not reach the server. Check the connection.');
    } finally {
      clearTimeout(t);
    }
    const json = await res.json().catch(() => ({}));
    if (res.status === 200 || res.status === 201) return Order.fromJson(json);
    const err = json.error;
    const message =
      err != null && typeof err === 'object' ? (err.message as string) ?? 'Checkout failed' : 'Checkout failed';
    throw new CheckoutException(message);
  }
}

/// Offline/demo checkout: builds an Order locally from the cart totals.
export class MockCheckoutRepository implements CheckoutRepository {
  async checkout(opts: { sessionId: string; token: string; cart: Cart; customer?: CustomerDraft }): Promise<Order> {
    const now = Date.now();
    return new Order({
      id: `demo_${now}`,
      status: 'placed',
      itemCount: opts.cart.count,
      subtotal: opts.cart.subtotal,
      tax: opts.cart.tax,
      total: opts.cart.total,
      createdAt: new Date(now).toISOString(),
      customerName: opts.customer?.name,
    });
  }
}
